File: benchmark/support/comparison.py
import sys
import time
from collections.abc import Callable
from statistics import median


def run_comparison(
    name: str,
    run: Callable[[int], int],
    *,
    unit: str,
    iterations: int,
    warmup_iterations: int,
    default_samples: int = 7,
) -> None:
    """Match comparison.dart's warmups, timing boundaries, and checksum."""
    count = int(sys.argv[1]) if len(sys.argv) > 1 else iterations
    samples = int(sys.argv[2]) if len(sys.argv) > 2 else default_samples
    if count < 1 or samples < 7:
        raise ValueError('Positive iterations and at least seven samples required')

    checksum = 0
    for _ in range(2):
        checksum += run(warmup_iterations)
    times = []
    expected = None
    for _ in range(samples):
        start = time.perf_counter()
        result = run(count)
        elapsed = (time.perf_counter() - start) * 1000
        if expected is not None and result != expected:
            raise RuntimeError(f'Result changed between samples: {expected} -> {result}')
        expected = result
        checksum += result
        times.append(elapsed)
    ordered = sorted(times)
    middle = median(ordered)
    print(f'{name} median_ms={middle:.3f} '
          f'min_ms={ordered[0]:.3f} max_ms={ordered[-1]:.3f} '
          f'ns/{unit}={middle * 1000000 / count:.2f} '
          f'raw_ms={",".join(f"{value:.3f}" for value in times)}')
    print(f'checksum={checksum}')

File: benchmark/support/comparison.dart
import 'package:dart_eval/dart_eval.dart';

/// Times guest execution after compilation, using the same sampling protocol
/// as comparison.py. Compile each benchmark host with `dart compile exe`.
void runComparison(
  List<String> args, {
  required String name,
  required String source,
  required String parameter,
  required String unit,
  required int iterations,
  required int warmupIterations,
}) {
  final count = args.isEmpty ? iterations : int.parse(args[0]);
  final samples = args.length < 2 ? 7 : int.parse(args[1]);
  if (count < 1 || samples < 7) {
    throw ArgumentError(
      'Positive iterations and at least seven samples required',
    );
  }
  final library = 'package:$name/main.dart';
  final compiler = Compiler()..entrypoints.add(library);
  final program = compiler.compile({
    name: {'main.dart': source},
  });
  final runtime = Runtime(program.write().buffer);
  int run(int n) =>
      runtime.executeLib(library, 'main', arguments: {parameter: n}) as int;

  var checksum = 0;
  for (var warm = 0; warm < 2; warm++) {
    checksum += run(warmupIterations);
  }
  final times = <double>[];
  int? expected;
  for (var sample = 0; sample < samples; sample++) {
    final watch = Stopwatch()..start();
    final result = run(count);
    watch.stop();
    if (expected != null && result != expected) {
      throw StateError('Result changed between samples: $expected -> $result');
    }
    expected = result;
    checksum += result;
    times.add(watch.elapsedMicroseconds / 1000);
  }
  final raw = List<double>.of(times);
  times.sort();
  final middle = times.length ~/ 2;
  final median = times.length.isOdd
      ? times[middle]
      : (times[middle - 1] + times[middle]) / 2;
  print(
    '$name median_ms=${median.toStringAsFixed(3)} '
    'min_ms=${times.first.toStringAsFixed(3)} '
    'max_ms=${times.last.toStringAsFixed(3)} '
    'ns/$unit=${(median * 1000000 / count).toStringAsFixed(2)} '
    'raw_ms=${raw.map((value) => value.toStringAsFixed(3)).join(',')}',
  );
  print('checksum=$checksum');
}
